### PR TITLE
[Parse] Several improvement for attribute parsing.

### PR DESCRIPTION
**Pull request by rintaro**
### [Parse] Eliminate Parser::canParseAttributes()

`@foo=bar` style attributes are no longer supported anyway.
So as `,` separated attribute list.

In `canParseTypeTupleBody()`, `canParseType()` can more accurately consume
type attributes.

In `isStartOfGetSetAccessor()`, we can trivially inline the functionality.
### [QoI][Parse] Improve recovering from unknown attribute

By consuming parens.
As for type attributes, handle `@unknownAttribute(Arg) -> Ret` case.

Improves diagnostic QoI. For example, on:
`func foo(x: @unknown(x) Int) {}`

Before:

```
test.swift:1:14: error: unknown attribute 'unknown'
test.swift:1:25: error: expected ',' separator
test.swift:1:25: error: unnamed parameters must be written with the empty name '_'
test.swift:1:22: error: use of undeclared type 'x'
```

Now, we just emit the first one:

```
test.swift:1:14: error: unknown attribute 'unknown'
func foo(x: @unknown(x) Int) {}
             ^
```
### [Parse] Remove unnecessary parameters from `parseDeclAttributeList` function.

`InParam` was not used at all.

`StopAtTypeAttributes`
As far as I understand, this option _was_ merely for improving diagnostic QoI
for declarations like:
  `func foo(@typeattr Arg) {}`
to fix-it to:
  `func foo(_: @typeattr Arg) {}`

But, this causes the very loudly diagnostics for misplaced type attributes.
For example, on:
`func foo(@convention(block) x: () -> CInt) {}`

```
test.swift:1:10: error: expected parameter name followed by ':'
test.swift:1:10: error: expected ',' separator
test.swift:1:10: error: expected ')' in parameter
test.swift:1:9: note: to match this opening '('
test.swift:1:10: error: consecutive statements on a line must be separated by ';'
test.swift:1:11: error: attribute can only be applied to types, not declarations
test.swift:1:21: error: expected declaration
test.swift:1:44: error: statement cannot begin with a closure expression
test.swift:1:44: note: explicitly discard the result of the closure by assigning to '_'
test.swift:1:44: error: braced block of statements is an unused closure
test.swift:1:6: error: expected '{' in body of function declaration
test.swift:1:44: error: expression resolves to an unused function
```

Now, we emit more accurate diagnostic:

```
test.swift:1:11: error: attribute can only be applied to types, not declarations
func foo(@convention(block) x: () -> CInt) {}
          ^
```

Note that This causes small regression in diagnostics for bare type parameter
like `func foo(@convention(c) () -> CInt) {}`:

Before:

```
test.swift:1:10: error: unnamed parameters must be written with the empty name '_'
func foo(@convention(block) () -> CInt) {}
         ^
         _:
```

Now:

```
test.swift:1:11: error: attribute can only be applied to types, not declarations
func foo(@convention(block) () -> CInt) {}
          ^
test.swift:1:29: error: unnamed parameters must be written with the empty name '_'
func foo(@convention(block) () -> CInt) {}
                            ^
                            _:
```
### [AST] Remove @escaping from decl attributes.

`@escaping` is not really a decl attribute, ever since it was introduced.

### CODE REVIEW

**Review by jrose-apple — CHANGES_REQUESTED** [review]

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -1312,15 +1315,14 @@ bool Parser::parseTypeAttribute(TypeAttributes &Attributes, bool justChecking) {
...
-    if (consumeIf(tok::equal)) {
-      if (Tok.is(tok::identifier) ||
-          Tok.is(tok::integer_literal) ||
-          Tok.is(tok::floating_literal))
-        consumeToken();
+    if(Tok.is(tok::l_paren)) {
+      ParserPosition LParenPosition = getParserPosition();
+      skipSingle();
+      // If we found '->' after paren, it's likely a parameter of function type.
+      if (Tok.is(tok::arrow))
```
Comment (rintaro) [inline lib/Parse/ParseDecl.cpp]: Thanks! Added `getEndOfPreviousLoc() == Tok.getLoc()` condition, and `throws` after paren check.
How do you think now?

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Review by jrose-apple — APPROVED** [review]
Looks great. Nice cleanup, thanks Rintaro!

**Comment by rintaro** [comment]
Thank you for reviewing!

**Pull request merged**